src/raftstore/store/worker/scan.rs

```diff
@@ -0,0 +1,355 @@
+        let region_id = region.get_id();
+        info!("[region {}] computing hash at {}", region_id, index);
+        REGION_HASH_COUNTER_VEC.with_label_values(&["compute", "all"]).inc();
+
+        let timer = REGION_HASH_HISTOGRAM.start_timer();
+        let mut sha1 = Sha1::new();
+        let mut cf_names = snap.cf_names();
+        cf_names.sort();
+        let start_key = keys::enc_start_key(&region);
+        let end_key = keys::enc_end_key(&region);
+        for cf in cf_names {
+            let res = snap.scan_cf(cf,
+                                   &start_key,
+                                   &end_key,
+                                   false,
+                                   &mut |k, v| {
+                                       update(&mut sha1, k);
+                                       update(&mut sha1, v);
+                                       Ok(true)
+                                   });
+            if let Err(e) = res {
+                REGION_HASH_COUNTER_VEC.with_label_values(&["compute", "failed"]).inc();
+                error!("[region {}] failed to calculate hash: {:?}", region_id, e);
+                return;
+            }
+        }
+        let region_state_key = keys::region_state_key(region_id);
+        update(&mut sha1, &region_state_key);
+        match snap.get_value_cf(CF_RAFT, &region_state_key) {
+            Err(e) => {
+                REGION_HASH_COUNTER_VEC.with_label_values(&["compute", "failed"]).inc();
+                error!("[region {}] failed to get region state: {:?}", region_id, e);
+                return;
+            }
+            Ok(Some(v)) => update(&mut sha1, &v),
+            Ok(None) => update(&mut sha1, b""),
+        }
+        let digest = sha1.digest();
+        timer.observe_duration();
+
+        let msg = Msg::ComputeHashResult {
+            region_id: region_id,
+            index: index,
+            hash: digest.bytes().to_vec(),
+        };
+        self.checksum.insert(region_id, (index, digest));
+        if let Err(e) = self.ch.try_send(msg) {
+            warn!("[region {}] failed to send hash compute result, err {:?}",
+                  region_id,
+                  e);
+        }
+    }
+
+    fn verify_hash(&mut self, region_id: u64, expect_index: u64, expect_hash: Vec<u8>) {
+        if let Some(&(index, ref digest)) = self.checksum.get(&region_id) {
+            if index != expect_index {
+                REGION_HASH_COUNTER_VEC.with_label_values(&["verify", "miss"]).inc();
+                warn!("[region {}] computed hash belongs to index {}, but we want {}, skip.",
+                      region_id,
+                      index,
+                      expect_index);
+                return;
+            }
+            let actual_hash = digest.bytes();
+            if actual_hash != expect_hash.as_slice() {
+                REGION_HASH_COUNTER_VEC.with_label_values(&["verify", "failed"]).inc();
```

Comment:
seem that it is not necessary to add metric here, we will panic immediately later.